File: data-structures-sorting-algorithms-advanced/src/algorithms/quickSort.js
```javascript
/**
 * @fileoverview Implementation of the Quick Sort algorithm using Lomuto partition scheme.
 */

const { swap } = require('../utils/arrayUtils');

/**
 * Partitions the array around a pivot element using Lomuto partition scheme.
 * Elements smaller than or equal to the pivot are moved to its left,
 * and elements greater than the pivot are moved to its right.
 * The pivot element ends up in its final sorted position.
 *
 * @param {Array<number>} arr The array to partition.
 * @param {number} low The starting index of the subarray to partition.
 * @param {number} high The ending index of the subarray to partition.
 * @returns {number} The final index of the pivot element.
 */
function partition(arr, low, high) {
  // Choose the last element as the pivot.
  // Other pivot strategies exist: first element, middle element, random element.
  const pivot = arr[high];
  let i = low - 1; // Index of smaller element

  for (let j = low; j < high; j++) {
    // If current element is smaller than or equal to pivot
    if (arr[j] <= pivot) {
      i++; // Increment index of smaller element
      swap(arr, i, j);
    }
  }

  // Place the pivot element at its correct position
  swap(arr, i + 1, high);
  return i + 1; // Return the pivot's final index
}

/**
 * Recursive helper function for Quick Sort.
 *
 * @param {Array<number>} arr The array to be sorted.
 * @param {number} low The starting index of the subarray to sort.
 * @param {number} high The ending index of the subarray to sort.
 * @returns {void} The array is modified in-place.
 */
function _quickSort(arr, low, high) {
  if (low < high) {
    // pi is partitioning index, arr[pi] is now at right place
    const pi = partition(arr, low, high);

    // Recursively sort elements before partition and after partition
    _quickSort(arr, low, pi - 1);
    _quickSort(arr, pi + 1, high);
  }
}

/**
 * Sorts an array of numbers using the Quick Sort algorithm.
 *
 * Quick Sort is a divide-and-conquer algorithm. It works by:
 * 1. Picking an element as a pivot.
 * 2. Partitioning the array around the pivot: all elements smaller than the pivot
 *    come before it, and all elements larger come after it. The pivot is now
 *    in its final sorted position.
 * 3. Recursively applying the above steps to the subarrays formed by the partition.
 *
 * Characteristics:
 * - Highly efficient in practice, often faster than Merge Sort and Heap Sort.
 * - Unstable: Yes (does not preserve the relative order of equal elements).
 * - In-place: Yes (most implementations are, especially with Lomuto or Hoare partition).
 * - Not adaptive: Performance doesn't significantly improve on already sorted data (can even degrade).
 *
 * @param {Array<number>} arr The array of numbers to be sorted.
 * @returns {Array<number>} The sorted array. The original array is modified in-place.
 *
 * Time Complexity:
 *   - Worst-case: O(N^2) (occurs when the pivot selection consistently leads to
 *                     unbalanced partitions, e.g., already sorted array with
 *                     last element as pivot).
 *   - Average-case: O(N log N)
 *   - Best-case: O(N log N)
 *
 * Space Complexity:
 *   - O(log N) on average (due to recursion stack for balanced partitions).
 *   - O(N) in worst-case (for highly unbalanced partitions).
 *
 * Example:
 *   let arr = [5, 1, 4, 2, 8];
 *   quickSort(arr); // arr becomes [1, 2, 4, 5, 8]
 *   quickSort([]);   // returns []
 *   quickSort([1]);  // returns [1]
 */
function quickSort(arr) {
  if (!Array.isArray(arr)) {
    throw new Error('Input must be an array.');
  }
  const n = arr.length;

  // An array with 0 or 1 element is already sorted.
  if (n <= 1) {
    return arr;
  }

  // Call the recursive helper function
  _quickSort(arr, 0, n - 1);
  return arr;
}

module.exports = quickSort;
```

File: data-structures-sorting-algorithms-enhanced/sorting_algorithms_project/algorithms/quick_sort.py
import random

def quick_sort(arr, pivot_choice="last"):
    """
    Sorts an array using the Quick Sort algorithm.

    Quick Sort is a highly efficient, in-place, comparison-based sorting algorithm.
    It works by selecting a 'pivot' element from the array and partitioning the
    other elements into two sub-arrays, according to whether they are less than
    or greater than the pivot. The sub-arrays are then sorted recursively.

    Args:
        arr (list): The list of comparable elements to be sorted.
        pivot_choice (str): Strategy for choosing the pivot.
                            Options: "last", "first", "middle", "random".
                            Default is "last".

    Time Complexity:
        - Worst Case: O(n^2) - Occurs when the pivot selection consistently results in
                               highly unbalanced partitions (e.g., already sorted array
                               with last element as pivot, or all elements are same).
        - Average Case: O(n log n) - Achieved with good pivot selection that creates
                                     balanced partitions.
        - Best Case: O(n log n) - When the pivot consistently divides the array into
                                  two equal halves.

    Space Complexity:
        - O(log n) - Average case for the recursive call stack.
        - O(n) - Worst case for the recursive call stack (unbalanced partitions).
                 This is generally considered "in-place" because the auxiliary space
                 is for the call stack, not for storing duplicates of the data.

    Stability:
        - Unstable: Quick Sort is generally not stable. The relative order of equal
                    elements may not be preserved due to swapping elements far apart
                    during partitioning.
    """
    _quick_sort_recursive(arr, 0, len(arr) - 1, pivot_choice)

def _quick_sort_recursive(arr, low, high, pivot_choice):
    """
    Recursive helper function for Quick Sort.
    """
    if low < high:
        # pi is partitioning index, arr[pi] is now at right place
        pi = _partition(arr, low, high, pivot_choice)

        # Separately sort elements before partition and after partition
        _quick_sort_recursive(arr, low, pi - 1, pivot_choice)
        _quick_sort_recursive(arr, pi + 1, high, pivot_choice)

def _partition(arr, low, high, pivot_choice):
    """
    Partition scheme (Lomuto Partition Scheme is commonly used).
    It places the pivot element at its correct sorted position and places
    all smaller elements to its left and all greater elements to its right.
    """
    # 1. Choose Pivot
    if pivot_choice == "first":
        pivot_idx = low
    elif pivot_choice == "middle":
        pivot_idx = (low + high) // 2
    elif pivot_choice == "random":
        pivot_idx = random.randint(low, high)
    else: # Default and "last"
        pivot_idx = high

    pivot_value = arr[pivot_idx]

    # Move pivot to the end (if not already there) to simplify partition logic
    # This prepares the array for Lomuto's partition scheme where pivot is at the end.
    if pivot_idx != high:
        arr[pivot_idx], arr[high] = arr[high], arr[pivot_idx]
    # Now the pivot is arr[high]

    i = low - 1  # Index of smaller element

    for j in range(low, high):
        # If current element is smaller than or equal to pivot
        if arr[j] <= pivot_value:
            # Increment index of smaller element and swap
            i += 1
            arr[i], arr[j] = arr[j], arr[i]

    # Swap the pivot element (currently at arr[high]) with the element at i + 1
    # This places the pivot in its correct sorted position
    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1 # Return the index of the pivot element

# Alternative Partition Scheme (Hoare's Partition Scheme)
# Hoare's scheme is generally more efficient than Lomuto's because it does fewer swaps
# on average, especially for arrays with many duplicate elements.
# It places the pivot such that elements before it are less than or equal to,
# and elements after it are greater than or equal to the pivot.
def _hoare_partition(arr, low, high, pivot_choice):
    """
    Hoare's Partition Scheme.
    """
    # 1. Choose Pivot
    if pivot_choice == "first":
        pivot_idx = low
    elif pivot_choice == "middle":
        pivot_idx = (low + high) // 2
    elif pivot_choice == "random":
        pivot_idx = random.randint(low, high)
    else: # Default and "last"
        pivot_idx = high

    pivot_value = arr[pivot_idx]
    # Move pivot to the beginning
    arr[pivot_idx], arr[low] = arr[low], arr[pivot_idx]
    # Now pivot is arr[low]

    i = low - 1
    j = high + 1

    while True:
        i += 1
        while arr[i] < pivot_value:
            i += 1

        j -= 1
        while arr[j] > pivot_value:
            j -= 1

        if i >= j:
            return j # Return the split point

        arr[i], arr[j] = arr[j], arr[i]

# To use Hoare's partition:
# Change `pi = _partition(arr, low, high, pivot_choice)` to `pi = _hoare_partition(arr, low, high, pivot_choice)`
# And change recursive calls to:
# `_quick_sort_recursive(arr, low, pi, pivot_choice)`
# `_quick_sort_recursive(arr, pi + 1, high, pivot_choice)`
```